### PR TITLE
Through Install node collect logs from other host

### PR DESCRIPTION
**Pull request by anrao19**
Signed-off-by: Anuchaithra Rao <anrao@redhat.com>

# Description
Through Install node collect logs from other host. Then push it to Magna
http://magna002.ceph.redhat.com/cephci-jenkins/ceph_logs/

latest logs:
http://magna002.ceph.redhat.com/cephci-jenkins/ceph_logs_AOH9OF/

# Checklist:

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarin Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test

### CODE REVIEW

File: collect_logs.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        ssh_d.exec_command(f"sudo rm -rf {sosreport.group()}")
+        ssh_d.close()
+        print(f"Successfully copied report to {nodeip}")
+    except Exception:
+        results[nodeip] = False
+
+
+def run(args):
+    """
+    Using the provided installer ip get all other nodes then generate sosreport for all the nodes obtained.
```
Comment (psathyan) [inline collect_logs.py:107]: Please correct the docString

File: collect_logs.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    Run shell script on installer node based on its ceph version, then upload all the collected logs to magna
+    :param:
+       - ip        :installer IP address
+       - username  :Username other than root
+       - password  :password for installer node
+       - log-dir   :name of the output directory [default tmp]
```
Comment (psathyan) [inline collect_logs.py:113]: IMO, we could do the docString like below... it helps render the documentation properly with readthedocs
```
Args:
  ip                  Node IP address or FQDN
  username     the username to be used to access the system
  log-dir           the place to store the collected logs
```
`log-dir`, I thought we were planning to use `magna002`

File: collect_logs.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    Returns:
+        0 on success or 1 for failures
+    """
+    results = dict()
+    run_id = generate_unique_id(length=6)
+    ip = args["--ip"]
+    uname = args["--username"]
+    pword = args["--password"]
+    log_dir = f"ceph_logs_{run_id}"
+    magna = "http://magna002.ceph.redhat.com/cephci-jenkins"
```
Comment (psathyan) [inline collect_logs.py:123]: Instead of `magna`, we could use the variable `file_share`

File: collect_logs.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        ssh_install.exec_command(f"sudo mkdir -p {log_dir}")
+        ssh_install.exec_command(f"sudo chown -R {uname}:{uname} {log_dir}")
+        stdin, stdout, stderr = ssh_install.exec_command(
+            "cut -f 1 /etc/hosts | cut -d ' ' -f 1"
+        )
+        nodes = stdout.read().decode().split("\n")
+        print(f"Host that are obtained from given host: {nodes}")
+        with parallel() as p:
+            for nodeip in nodes:
+                if nodeip.startswith("10.0"):
```
Comment (psathyan) [inline collect_logs.py:137]: Could we have a way to exclude `localhost` instead of matching with `ip address`?

File: collect_logs.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                        generate_sosreport_in_node,
+                        ssh_install,
+                        nodeip,
+                        uname,
+                        pword,
+                        log_dir,
+                        results,
+                    )
+        collect_logs(ip, ssh_install, uname, log_dir, results)
+        upload_logs(log_dir, ssh_install, nodeip, results)
```
Comment (psathyan) [inline collect_logs.py:148]: Can do these steps in parallel?

File: collect_logs.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    run_id = generate_unique_id(length=6)
+    ip = args["--ip"]
+    uname = args["--username"]
+    pword = args["--password"]
+    log_dir = f"ceph_logs_{run_id}"
+    magna = "http://magna002.ceph.redhat.com/cephci-jenkins"
+
+    ssh_install = SSHConnectionManager(ip, uname, pword).get_client()
+    stdin, stdout, stderr = ssh_install.exec_command("hostname")
+    if "installer" in stdout.read().decode():
```
Comment (psathyan) [inline collect_logs.py:127]: Can we do it like
```
if "installer" not in stdout.read().decode():
    raise AssertionError

```

File: collect_logs.py (hunk 5 above)
Comment (anrao19) [reply collect_logs.py:148]: This operation will be performed on single node 
so parallel is not required

File: collect_logs.py (hunk 6 above)
Comment (anrao19) [reply collect_logs.py]: Made changes and recommitted

File: collect_logs.py (hunk 1 above)
Comment (anrao19) [reply collect_logs.py]: Made changes

**Comment by rakeshgm** [comment]
@anrao19 can you also collect audit logs, syslogs as well  and kernel logs. it helps i believe

File: collect_logs.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    Raises:
+        AssertionError: An error occurred if given IP is not of Installer node
+    """
+    results = dict()
+    run_id = generate_unique_id(length=6)
+    ip = args["--ip"]
+    uname = args["--username"]
+    pword = args["--password"]
+    log_dir = f"ceph_logs_{run_id}"
+    file_share = "http://magna002.ceph.redhat.com/cephci-jenkins"
```
Comment (rakeshgm) [inline collect_logs.py:148]: I was thinking that we should move the logs location under `ceph-jenkins/ceph-<run_id>/ceph_logs`

**Comment by anrao19** [comment]
> @anrao19 can you also collect audit logs, syslogs as well and kernel logs. it helps i believe

@rakeshgm these logs you can find in the sosreport generated

File: collect_logs.py (hunk 7 above)
Comment (anrao19) [reply collect_logs.py:150]: Changed Location to cephci-jenkins/ceph_logs_{run_id}/logs

**Comment by rakeshgm** [comment]
we have discussed couple of more changes, @anrao19 will try if they can be added into this.

**Comment by anrao19** [comment]
> we have discussed couple of more changes, @anrao19 will try if they can be added into this.

@rakeshgm  made changes as we discussed.

**Comment by rakeshgm** [comment]
@anrao19 looks good.

**Review by rakeshgm — APPROVED** [review]

**Review by psathyan — APPROVED** [review]

**Pull request merged**